opentargets_pharmgkb/evidence_generation.py

```diff
@@ -80,72 +84,117 @@ def pipeline(data_dir, fasta_path, created_date, output_path, debug_path=None):
+    return f'{chrom}_{pos}_{ref}_{",".join(parsed_genotype)}' if chrom and pos and ref else None
+
+
+def get_genotype_ids(df, fasta_path, counts=None):
     """
-    Get VCF-style coordinates (chr_pos_ref_alt) for dataframe.
+    Get genotype IDs (chr_pos_ref_allele1,allele2) for dataframe.
 
-    :param df: dataframe to annotate (needs 'Genotype/Allele' and 'Variant/Haplotypes' columns)
-    :return: dataframe with 'vcf_coords' column added
+    :param df: dataframe to annotate (needs 'Genotype/Allele', 'Variant/Haplotypes', 'Location' columns)
+    :param fasta_path: path to fasta file to check reference
+    :param counts: ClinicalAnnotationCounts; if provided will count multi-allelic variants.
+    :return: dataframe with 'genotype_id' column added
     """
     fasta = Fasta(fasta_path)
-    # First set a column with all genotypes for a given rs
-    df_with_coords = pd.merge(df, df.groupby(by=ID_COL_NAME).aggregate(
-        all_genotypes=('Genotype/Allele', list)), on=ID_COL_NAME)
-    # Then get coordinates for each row
-    # TODO Currently this does one call per genotype per RS
-    #  Remove redundant calls once we figure out how to handle genotypes & multiple alts per RS
-    for i, row in df_with_coords.iterrows():
-        df_with_coords.at[i, 'vcf_coords'] = fasta.get_coordinates_for_clinical_annotation(
-            row['Variant/Haplotypes'], row['Location'], row['all_genotypes'])
-    return df_with_coords
+    # First set a column with all genotypes for a given RS
+    df_with_ids = df.assign(parsed_genotype=df['Genotype/Allele'].apply(parse_genotype))
+    df_with_ids = pd.merge(df_with_ids, df_with_ids.groupby(by='Variant/Haplotypes').aggregate(
+        all_genotypes=('parsed_genotype', list)), on='Variant/Haplotypes')
+    # Get coordinates for each RS
+    rs_to_coords = {}
+    for i, row in df_with_ids.drop_duplicates(['Variant/Haplotypes']).iterrows():
+        chrom, pos, ref, alleles_dict = fasta.get_chr_pos_ref(row['Variant/Haplotypes'], row['Location'],
+                                                              row['all_genotypes'])
+        rs_to_coords[row['Variant/Haplotypes']] = (chrom, pos, ref, alleles_dict)
+        if not counts:
+            continue
+        counts.total_rs += 1
+        if not alleles_dict:
+            continue
+        counts.rs_with_alleles += 1
+        if len(alleles_dict) <= 2:
+            continue
+        counts.rs_with_more_than_2_alleles += 1
+    # Get ID for each genotype
+    for i, row in df_with_ids.iterrows():
+        chrom, pos, ref, alleles_dict = rs_to_coords[row['Variant/Haplotypes']]
+        if chrom and pos and ref and alleles_dict:
+            df_with_ids.at[i, 'genotype_id'] = genotype_id(chrom, pos, ref, sorted([alleles_dict[a]
+                                                                                  for a in row['parsed_genotype']]))
+        else:
+            df_with_ids.at[i, 'genotype_id'] = None
```

Comment:
Couldn't this bit be included in the upper loop, so that we only iterate once over ``df_with_ids``?
Not fully sure why we drop duplicates in one but not the other 🤔, but if ``df_with_ids`` is a pretty long df, it may be worth it to mix the iteration in some way.

Reply:
Roger that. It was just that I noticed similar variables were being used in both loops